### PR TITLE
feat: Separate fine-tuning backend from the transformers backend

### PR DESCRIPTION
**Pull request by MartinRepo**
## Description
This PR introduces a `ft_backend` specifically for handling fine-tuning jobs. The RoundRobinRouter is updated to distinguish fine-tuning requests from inference requests and launch fine-tuning instances via start_ft_instance instead of the shared inference backend.

## Motivation
In the alpha version, fine-tuning requests were naively routed to the same inference backend (i.e., Transformers-based) without accounting for the fundamentally different nature of fine-tuning workloads. This caused resource contention, lack of isolation, and unnecessary assumptions about backend capabilities.

This update decouples fine-tuning from inference logic by introducing a dedicated `ft_backend` path. Fine-tuning requests now spin up their own isolated backend instance via `start_ft_instance`, enabling better specialization, resource control, and lifecycle management.

### Submit a fine-tuning job
```
curl -X POST $LLM_SERVER_URL/v1/fine-tuning/jobs \
  -H "Content-Type: application/json" \
  -d @examples/fine_tuning/fine_tuning_config.json
```
### Expected output
```
{
"job_id": "job-123",
}
```
### Get job status
```
curl -X GET "$LLM_SERVER_URL/v1/fine_tuning/jobs/{fine_tuning_job_id}"
```
### Cancel the job
```
curl -X POST "$LLM_SERVER_URL/v1/fine_tuning/jobs/{fine_tuning_job_id}/cancel"
```

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## Checklist
- [x] I have read the [CONTRIBUTION](https://github.com/ServerlessLLM/ServerlessLLM/blob/main/CONTRIBUTING.md) guide.
- [x] I have updated the tests (if applicable).
- [x] I have updated the documentation (if applicable).

### CODE REVIEW

**Review by dalongbao — CHANGES_REQUESTED** [review]
Fine-tuning works well, perfectly fine, LGTM; however, failed jobs are also marked as complete. Please make sure failed jobs are marked as failed. 

Here I used a dataset with an incompatible format:
## CLI
### Command
```bash
curl -X GET "$LLM_SERVER_URL/get-job-status?job_id=65bb4144-87cc-47ff-ac10-11e926a41dc2"
```
### Response
```json
{
  "fine-tuning job 65bb4144-87cc-47ff-ac10-11e926a41dc2 status": {
    "config": {
      "model": "facebook/opt-125m",
      "ft_backend": "peft_lora",
      "num_gpus": 1,
      "num_cpus": 1,
      "timeout": 3600,
      "backend_config": {
        "output_dir": "facebook/adapters/opt-125m_adapter_test",
        "dataset_config": {
          "dataset_source": "hf_hub",
          "hf_dataset_name": "zhonglongbao/makise-kurisu-chat-lora",
          "tokenization_field": "prompt",
          "split": "train",
          "data_files": "",
          "extension_type": ""
        },
        "lora_config": {
          "r": 4,
          "lora_alpha": 32,
          "lora_dropout": 0.05,
          "bias": "none",
          "task_type": "CAUSAL_LM"
        },
        "training_config": {
          "auto_find_batch_size": true,
          "save_strategy": "no",
          "num_train_epochs": 2,
          "learning_rate": 0.0001,
          "use_cpu": false
        }
      }
    },
    "status": "completed", # Should be failed
    "created_time": "2025-07-29T18:30:03.050847",
    "updated_time": "2025-07-29T18:30:03.050852",
    "priority": 0
  }
}
```
## Docker logs
```bash
(PeftLoraBackend pid=941, ip=172.18.0.2) INFO 07-29 18:30:32 peft_lora_backend.py:249] Started fine-tuning job: 65bb4144-87cc-47ff-ac10-11e926a41dc2
(PeftLoraBackend pid=941, ip=172.18.0.2) INFO 07-29 18:30:32 peft_lora_backend.py:252] Loading dataset with config: {'dataset_source': 'hf_hub', 'hf_dataset_name': 'zhonglongbao/makise-kurisu-chat-lora', 'tokenization_field': 'prompt', 'split': 'train', 'data_files': '', 'extension_type': ''}
(RoundRobinRouter pid=484) INFO 07-29 18:30:34 roundrobin_router.py:245] Finished processing fine-tuning facebook/opt-125m
(RoundRobinRouter pid=484) INFO 07-29 18:30:34 roundrobin_router.py:567] Force deleting an instance (even if it is busy) for model facebook/opt-125m
(PeftLoraBackend pid=941, ip=172.18.0.2) ERROR 07-29 18:30:34 peft_lora_backend.py:259] Failed to load dataset: 'prompt'
(PeftLoraBackend pid=941, ip=172.18.0.2) INFO 07-29 18:30:34 peft_lora_backend.py:349] Shutting down fine-tuning backend
Map:   0%|          | 0/1708 [00:00<?, ? examples/s]
(SllmController pid=474) INFO 07-29 18:30:35 controller.py:428] [FT Job 65bb4144-87cc-47ff-ac10-11e926a41dc2] Fine-tuning completed successfully.
(SllmController pid=474) INFO 07-29 18:30:35 controller.py:434] [FT Job 65bb4144-87cc-47ff-ac10-11e926a41dc2] Job status updated to 'completed'.
(FcfsScheduler pid=480) INFO 07-29 18:30:35 fcfs_scheduler.py:92] Deallocating model facebook/opt-125m instance facebook/opt-125m_32e96f7a-6b52-4d69-a8aa-764358fea974
(FcfsScheduler pid=480) INFO 07-29 18:30:35 fcfs_scheduler.py:103] Node 0 deallocated 1 GPUs
(FcfsScheduler pid=480) INFO 07-29 18:30:35 fcfs_scheduler.py:108] Model facebook/opt-125m instance facebook/opt-125m_32e96f7a-6b52-4d69-a8aa-764358fea974 deallocated
(RoundRobinRouter pid=484) INFO 07-29 18:30:35 roundrobin_router.py:253] Fine-tuning completed successfully for model facebook/opt-125m
```

Status should be "failed". 

# System information:
- 1x A6000 GPU
- Ubuntu 22.04
- Python 3.10

Started via `docker-compose.yml`, using [this dataset](https://huggingface.co/datasets/zhonglongbao/makise-kurisu-chat-lora).

**Comment by dalongbao** [comment]
Also please update the documentation to reflect the changes.

**Comment by MartinRepo** [comment]
Hi @dalongbao Thanks for reporting the bug! I have fixed the status updating bug, and updated the documentation. Just let me know if there is any problem.

File: sllm/default_ft_config.json
[the remark this answers is not in the document]
Comment (MartinRepo) [reply sllm/default_ft_config.json:1]: I think it could be better if put it in `examples/fine_tuning/`, like `examples/fine_tuning/fine_tuning_config.json`. How do you think about it?

**Comment by future-xy** [comment]
I noticed two issues during end-to-end testing:

1. It looks like a pre-downloaded checkpoint is required for a fine-tuning request. Could you please double-check this? If that’s the case, I’d suggest we automatically download the model checkpoint to avoid requiring users to run `sllm-store download` manually.

2. During a fine-tuning request, the scheduler didn't register the model with the store, leading to the following error:
```
 *  Executing task: docker logs --tail 1000 -f 3f81799b57218fd064130686f60535bc35a4093acfd277f0c0f4c268e32834df 

Initializing worker node...
Activating worker conda environment...
No node IP specified. Ray will attempt to determine the best IP automatically.
Executing: ray start --address=sllm_head:6379 --resources='{"worker_node": 1, "worker_id_0": 1}'
2025-08-27 10:57:09,852 INFO scripts.py:1152 -- Local node IP: 172.18.0.3
2025-08-27 10:57:09,974 WARNING services.py:2159 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 67108864 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing '--shm-size=10.24gb' to 'docker run' (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.
2025-08-27 10:57:10,107 SUCC scripts.py:1168 -- --------------------
2025-08-27 10:57:10,107 SUCC scripts.py:1169 -- Ray runtime started.
2025-08-27 10:57:10,107 SUCC scripts.py:1170 -- --------------------
2025-08-27 10:57:10,107 INFO scripts.py:1172 -- To terminate the Ray runtime, run
2025-08-27 10:57:10,107 INFO scripts.py:1173 --   ray stop
Starting sllm-store with arguments: --storage-path=/models --mem-pool-size 4GB --registration-required true
INFO 08-27 10:57:14 cli.py:194] Starting gRPC server
INFO 08-27 10:57:14 server.py:40] StorageServicer: storage_path=/models, mem_pool_size=4294967296, num_thread=4, chunk_size=33554432, registration_required=True
WARNING: Logging before InitGoogleLogging() is written to STDERR
I20250827 10:57:14.268049     1 checkpoint_store.cpp:41] Number of GPUs: 1
I20250827 10:57:14.268071     1 checkpoint_store.cpp:43] I/O threads: 4, chunk size: 32MB
I20250827 10:57:14.268077     1 checkpoint_store.cpp:45] Storage path: "/models"
I20250827 10:57:15.059409     1 checkpoint_store.cpp:71] GPU 0 UUID: 2eea4b8b-d9b5-8b04-e426-c8d0617dc87a
I20250827 10:57:15.059471     1 pinned_memory_pool.cpp:29] Creating PinnedMemoryPool with 128 buffers of 33554432 bytes
I20250827 10:57:20.714128     1 checkpoint_store.cpp:83] Memory pool created with 4GB
INFO 08-27 10:57:20 server.py:231] Starting gRPC server on 0.0.0.0:8073
E20250827 10:58:29.609730   568 checkpoint_store.cpp:304] Model /models/transformers/facebook/opt-125m is not registered
E20250827 10:58:29.609759   568 checkpoint_store.cpp:115] Model /models/transformers/facebook/opt-125m is not registered
INFO 08-27 10:58:29 server.py:114] LoadModel: success /models/transformers/facebook/opt-125m with target 1
INFO 08-27 10:58:30 server.py:114] LoadModel: success /models/transformers/facebook/opt-125m with target 2
E20250827 10:58:30.766845   572 checkpoint_store.cpp:304] Model /models/transformers/facebook/opt-125m is not registered
E20250827 10:58:30.766896   572 checkpoint_store.cpp:203] Model /models/transformers/facebook/opt-125m is not registered
E20250827 10:58:30.781673     1 checkpoint_store.cpp:267] Model /models/transformers/facebook/opt-125m is not registered
INFO 08-27 10:58:30 server.py:141] Confirm model failed, retry 1
E20250827 10:58:30.831889     1 checkpoint_store.cpp:267] Model /models/transformers/facebook/opt-125m is not registered
INFO 08-27 10:58:30 server.py:141] Confirm model failed, retry 2
E20250827 10:58:30.882146     1 checkpoint_store.cpp:267] Model /models/transformers/facebook/opt-125m is not registered
INFO 08-27 10:58:30 server.py:141] Confirm model failed, retry 3
E20250827 10:58:30.932403     1 checkpoint_store.cpp:267] Model /models/transformers/facebook/opt-125m is not registered
INFO 08-27 10:58:30 server.py:141] Confirm model failed, retry 4
E20250827 10:58:30.982602     1 checkpoint_store.cpp:267] Model /models/transformers/facebook/opt-125m is not registered
INFO 08-27 10:58:30 server.py:141] Confirm model failed, retry 5
ERROR 08-27 10:58:31 server.py:145] Confirm model /models/transformers/facebook/opt-125m replica 6eb946f4-b39b-4f0e-9c2b-2c3fa637814a failed
```
This should be straightforward to fix if we follow the same path used for inference.
